java.io.NotSerializableException when using JavaSerializer in v5.1.0, v6.1.0

# Description
With the new configurable schema converter feature (#268, #269), the class `DefaultSchemaConverter` is instantiated by default as member variable `schemaConverter` in `AvroDataToCatalyst`. Even though `AvroDataToCatalyst` as a case class is serializable by default, serialization fails when using the `JavaSerializer` with the following error message

```
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task not serializable: java.io.NotSerializableException: za.co.absa.abris.avro.sql.DefaultSchemaConverter
Serialization stack:
	- object not serializable (class: za.co.absa.abris.avro.sql.DefaultSchemaConverter, value: za.co.absa.abris.avro.sql.DefaultSchemaConverter@1ce2ce83)
	- field (class: za.co.absa.abris.avro.sql.AvroDataToCatalyst, name: schemaConverter, type: interface za.co.absa.abris.avro.sql.SchemaConverter)
	- object (class za.co.absa.abris.avro.sql.AvroDataToCatalyst, from_avro(value#647, (readerSchema,{"type":"record","name":"e2etest","fields":[{"name":"field1","type":"string"},{"name":"field2","type":"int"}]})))
	- field (class: org.apache.spark.sql.catalyst.expressions.IsNotNull, name: child, type: class org.apache.spark.sql.catalyst.expressions.Expression)
	- object (class org.apache.spark.sql.catalyst.expressions.IsNotNull, isnotnull(from_avro(value#647, (readerSchema,{"type":"record","name":"e2etest","fields":[{"name":"field1","type":"string"},{"name":"field2","type":"int"}]}))))
	- field (class: org.apache.spark.sql.execution.FilterExec, name: condition, type: class org.apache.spark.sql.catalyst.expressions.Expression)
	- object (class org.apache.spark.sql.execution.FilterExec, Filter isnotnull(from_avro(value#647, (readerSchema,{"type":"record","name":"e2etest","fields":[{"name":"field1","type":"string"},{"name":"field2","type":"int"}]})))
```

# How to fix
~Add `Serializable` trait to `SchemaConverter` to trait.~
Make `schemaConverter` lazy

--- a/src/main/scala/za/co/absa/abris/avro/sql/AvroDataToCatalyst.scala
+++ b/src/main/scala/za/co/absa/abris/avro/sql/AvroDataToCatalyst.scala
@@ -40,7 +40,7 @@ private[abris] case class AvroDataToCatalyst(
   schemaRegistryConf: Option[Map[String,String]]
 ) extends UnaryExpression with ExpectsInputTypes {
 
-  private val schemaConverter = loadSchemaConverter(config.schemaConverter)
+  @transient private lazy val schemaConverter = loadSchemaConverter(config.schemaConverter)
 
   override def inputTypes: Seq[BinaryType.type] = Seq(BinaryType)
 

--- a/src/test/scala/za/co/absa/abris/avro/sql/AvroDataToCatalystSpec.scala
+++ b/src/test/scala/za/co/absa/abris/avro/sql/AvroDataToCatalystSpec.scala
@@ -103,7 +103,7 @@ class AvroDataToCatalystSpec extends AnyFlatSpec with Matchers with BeforeAndAft
       ))
       .withSchemaConverter("nonexistent")
 
-    val ex = intercept[ClassNotFoundException](from_avro(col("avroBytes"), fromAvroConfig))
+    val ex = intercept[ClassNotFoundException](from_avro(col("avroBytes"), fromAvroConfig).expr.dataType)
     ex.getMessage should include ("nonexistent")
   }
 

--- a/src/test/scala/za/co/absa/abris/avro/sql/AvroDataToCatalystSpec.scala
+++ b/src/test/scala/za/co/absa/abris/avro/sql/AvroDataToCatalystSpec.scala
@@ -16,6 +16,8 @@
 
 package za.co.absa.abris.avro.sql
 
+import org.apache.spark.SparkConf
+import org.apache.spark.serializer.{JavaSerializer, KryoSerializer}
 import org.apache.spark.sql.functions.col
 import org.apache.spark.sql.types.{IntegerType, LongType, StructField, StructType}
 import org.scalatest.BeforeAndAfterEach
@@ -103,5 +105,19 @@ class AvroDataToCatalystSpec extends AnyFlatSpec with Matchers with BeforeAndAft
 
     val ex = intercept[ClassNotFoundException](from_avro(col("avroBytes"), fromAvroConfig))
     ex.getMessage should include ("nonexistent")
+  }
+
+  it should "be serializable" in {
+    val schemaString = TestSchemas.NATIVE_SIMPLE_NESTED_SCHEMA
+    val config = FromAvroConfig().withReaderSchema(schemaString)
+    val avroDataToCatalyst = from_avro(col("col"), config).expr
+
+    val javaSerializer = new JavaSerializer(new SparkConf())
+    javaSerializer.newInstance().serialize(avroDataToCatalyst)
+
+    val kryoSerializer = new KryoSerializer(new SparkConf())
+    kryoSerializer.newInstance().serialize(avroDataToCatalyst)
+
+    // test successful if no exception is thrown
   }
 }

--- a/src/test/scala/za/co/absa/abris/avro/sql/CatalystDataToAvroSpec.scala
+++ b/src/test/scala/za/co/absa/abris/avro/sql/CatalystDataToAvroSpec.scala
@@ -1,0 +1,49 @@
+/*
+ * Copyright 2022 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.abris.avro.sql
+
+import org.apache.avro.SchemaBuilder
+import org.apache.spark.SparkConf
+import org.apache.spark.serializer.{JavaSerializer, KryoSerializer}
+import org.apache.spark.sql.functions.col
+import org.scalatest.BeforeAndAfterEach
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+import za.co.absa.abris.avro.functions._
+import za.co.absa.abris.config.ToAvroConfig
+
+class CatalystDataToAvroSpec extends AnyFlatSpec with Matchers with BeforeAndAfterEach {
+  it should "be serializable" in {
+    val schema = SchemaBuilder
+      .record("foo")
+      .namespace("test_namespace")
+      .fields()
+      .name("int").`type`().intType().noDefault()
+      .endRecord()
+      .toString
+    val config = ToAvroConfig().withSchema(schema)
+    val catalystDataToAvro = to_avro(col("col"), config).expr
+
+    val javaSerializer = new JavaSerializer(new SparkConf())
+    javaSerializer.newInstance().serialize(catalystDataToAvro)
+
+    val kryoSerializer = new KryoSerializer(new SparkConf())
+    kryoSerializer.newInstance().serialize(catalystDataToAvro)
+
+    // test successful if no exception is thrown
+  }
+}

